prometheus/prometheus.go

```diff
@@ -107,3 +164,98 @@ func (obj *Prometheus) UpdatePgraphStartTime() error {
 	obj.pgraphStartTimeSeconds.SetToCurrentTime()
 	return nil
 }
+
+// AddManagedResource increments the Managed Resource counter and updates the resource status.
+func (obj *Prometheus) AddManagedResource(resUuid string, rtype string) error {
+	if obj == nil {
+		return nil // happens when mgmt is launched without --prometheus
+	}
+	obj.managedResources.With(prometheus.Labels{"kind": rtype}).Inc()
+	if err := obj.UpdateState(resUuid, rtype, ResStateOK); err != nil {
+		return errwrap.Wrapf(err, "Can't update the resource status in the map!")
+	}
+	return nil
+}
+
+// RemoveManagedResource decrements the Managed Resource counter and updates the resource status.
+func (obj *Prometheus) RemoveManagedResource(resUuid string, rtype string) error {
+	if obj == nil {
+		return nil // happens when mgmt is launched without --prometheus
+	}
+	obj.managedResources.With(prometheus.Labels{"kind": rtype}).Dec()
+	if err := obj.deleteState(resUuid); err != nil {
+		return errwrap.Wrapf(err, "Can't remove the resource status from the map!")
+	}
+	return nil
+}
+
+// deleteState removes the resources for the state map and re-populates the failing gauge.
+func (obj *Prometheus) deleteState(resUuid string) error {
+	if obj == nil {
+		return nil // happens when mgmt is launched without --prometheus
+	}
+	delete(obj.resourcesState, resUuid)
+	if err := obj.updateFailingGauge(); err != nil {
+		return errwrap.Wrapf(err, "Can't update the failing gauge!")
+	}
+	return nil
+}
+
+// UpdateState updates the state of the resources in our internal state map
+// then triggers a refresh of the failing gauge.
+func (obj *Prometheus) UpdateState(resUuid string, rtype string, newState ResState) error {
+	if obj == nil {
+		return nil // happens when mgmt is launched without --prometheus
+	}
+	obj.resourcesState[resUuid] = resStateWithKind{state: newState, kind: rtype}
+	if newState != ResStateOK {
+		var strState string
+		if newState == ResStateSoftFail {
+			strState = "soft"
+		} else if newState == ResStateHardFail {
+			strState = "hard"
+		} else {
+			return errors.New("State should be Soft or Hard failure!")
+		}
+		obj.failedResourcesTotal.With(prometheus.Labels{"kind": rtype, "failure": strState}).Inc()
+	}
+	if err := obj.updateFailingGauge(); err != nil {
+		return errwrap.Wrapf(err, "Can't update the failing gauge!")
+	}
+	return nil
+}
+
+// updateFailingGauge refreshes the failing gauge by parsking the internal
+// state map.
+func (obj *Prometheus) updateFailingGauge() error {
+	if obj == nil {
+		return nil // happens when mgmt is launched without --prometheus
+	}
+	var softFails, hardFails map[string]float64
+	softFails = make(map[string]float64)
+	hardFails = make(map[string]float64)
+	for _, v := range obj.resourcesState {
+		if v.state == ResStateSoftFail {
+			if _, ok := softFails[v.kind]; ok {
```

Comment:
if there is an error because softfailsv.kind does not exist, it means that ok will be set, in that case we want to set the counter to 0